File: problems/0113_path_sum/solution.js

// Definition for a binary tree node.
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function(root, targetSum) {

    const result = [];

    var preOrder = (root, target, curr=[]) => {
        console.log(curr)
        // base condition
        if (!root) return
        if (target === root.val && !root.left && !root.right) {
            result.push([...curr, root.val])
            return
        }
        // general condition
        curr.push(root.val)
        preOrder(root.left, target - root.val, curr)
        preOrder(root.right, target - root.val, curr)
        curr.pop()
    }

    preOrder(root, targetSum)

    return result;
};

/**
 * DFS => stack
 * BFS => queue
 */
var pathSum = function(root, targetSum) {
    if (!root) return []
    const result = [];

    let queue = [[root, [root.val], root.val]]
    while(queue.length) {
        const [root, tmpResult, sum] = queue.shift()
        if (sum === targetSum && !root.left && !root.right) {
            result.push(tmpResult)
        }
        if (root.left) {
            queue.push([root.left, [...tmpResult, root.left.val], sum + root.left.val])
        }
        if (root.right) {
            queue.push([root.right, [...tmpResult, root.right.val], sum + root.right.val])
        }
    }

    return result
};
var pathSum = function(root, sum) {
    if (!root) return []
    const results = []
    const queue = [[root, [root.val], root.val]]
    while (queue.length) {
        const [node, arr, cur] = queue.shift()
        if (!node.left && !node.right && cur === sum) {
            results.push(arr)
        }
        if (node.left) {
            queue.push([node.left, [...arr, node.left.val], cur + node.left.val])
        } 
        if (node.right) {
            queue.push([node.right, [...arr, node.right.val], cur + node.right.val])
        }
    }
    return results
}



let node1 = new TreeNode(1)
let node2 = new TreeNode(2)
let node3 = new TreeNode(3)
node1.left = node2
node1.right = node3

console.log(pathSum(node1, 3))

File: problems/0113_path_sum/solution.py
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution1:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        """此題應該很類似題 257. Binary Tree Paths 的概念，不過是
        要在 leaf node 檢查是否加總為 targetSum，並且 path 為一陣列而不是 str。
        若沒有符合條件則需回傳空陣列。
        """
        res = []
        def backtrack(node: Optional[TreeNode], path: List[int]):
            nonlocal res
            # return case (leaf node and sum(path) == targetSum)
            if (not node.left
                and not node.right
                and sum(path) == targetSum
            ):
                res.append(path[:])
                return
            # general case
            if left := node.left:
                backtrack(left, path + [left.val])
            if right := node.right:
                backtrack(right, path + [right.val])
        backtrack(root, [root.val])
        return res

class Solution2:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        """使用 stack 解決問題，在 stack 中同時儲存 (node, path, curr_sum)"""
        if not root: return 
        res = []
        stack = [(root, [root.val], root.val)]
        while stack:
            node, path, tmp = stack.pop()
            if not node.left and not node.right and tmp == targetSum:
                res.append(path[:])
            if left := node.left:
                stack.append((left, path+[left.val], tmp+left.val))
            if right := node.right:
                stack.append((right, path+[right.val], tmp+right.val))
        return res


if __name__ == '__main__':
    pass